core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -536,6 +548,21 @@ State drained(Substream substream) {
       return new State(buffer, drainedSubstreams, winningSubstream, cancelled, passThrough);
     }
 
+    /** The given substream is closed. */
+    @CheckReturnValue
+    @GuardedBy("lock")
+    State closed(Substream substream) {
+      substream.closed = true;
+      if (this.drainedSubstreams.contains(substream)) {
+        Set<Substream> drainedSubstreams = new HashSet<Substream>();
```

Comment:
Use the `HashSet(Collection)` copy constructor.

Reply:
Done.